mountpoint-s3/src/fs.rs

```diff
@@ -58,8 +58,39 @@ where
 {
     inode: Inode,
     full_key: String,
-    object_size: u64,
-    typ: FileHandleType<Client, Prefetcher>,
+    state: FileHandleState<Client, Prefetcher>,
+}
+
+enum FileHandleState<Client, Prefetcher>
```

Comment:
Not a blocker, but why the 2 layers of enums? It seems that in most (all?) cases where we need to match on one enum we also immediately have to match on the other as well. We could "inline" the `FileHandleType`and replace `Assigned` with its 2 cases.

Reply:
It's just easier to read in my opinion. Previously, the read handle can be an option which represents created but not yet started, now we also want to do the same with the write handle. It's true that we can add new variants of `FileHandleType` to represent `None` cases or `Closed` handles, but I think it's clearer if we represent them in a new enum and leave the `FileHandleType` to be only read or write.

Reply:
Sorry, maybe I wasn't too clear. I'm thinking something like this (with better names, possibly!):

```rust
enum FileHandleState<Client, Prefetcher>
where
    Client: ObjectClient + Send + Sync + 'static,
    Prefetcher: Prefetch,
{
    Created { lookup: LookedUp, flags: i32, pid: u32 },
    Read {
        request: Prefetcher::PrefetchResult<Client>,
        pid: u32,
    },
    Write(UploadState<Client>),
    Closed,
}
```

Definitely not as important as the mutex issue, though.

Reply:
Got it, I see what you mean now.